### PR TITLE
Allow get_config to return the same config more than once

### PR DESCRIPTION
**Pull request by mseeger**
*Issue #, if available:* 415

*Description of changes:*
Introduces flag `allow_duplicates` to searchers, which allows to return the same config more than once. Also contains a new test on searchers, whether they properly implement `allow_duplicates=False` (the default).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py [hunk 1]
```diff
@@ -122,6 +127,11 @@ def __init__(
 
         self._hp_ranges = make_hyperparameter_ranges(self.config_space)
         self._excl_list = ExclusionList.empty_list(self._hp_ranges)
+        self._allow_duplicates = allow_duplicates
+        # Maps ``trial_id`` to configuration. This is used to blacklist
```
Comment (mseeger) [inline syne_tune/optimizer/schedulers/searchers/bore/bore.py:131]: This is a technicality:
- If not `allow_duplicates`, then `_excl_list` collects all configs suggested, so none is suggested 2x
- If `allow_duplicates`, then `_excl_list` only collects configs for which the trial failed. In this case, we need to know the config for a trial, because `evaluation_failed` only gets the `trial_id`.


The same code is used in other places, same comment applies.

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py [hunk 2]
```diff
@@ -293,5 +304,24 @@ def _update(self, trial_id: str, config: dict, result: dict):
...
+                self._config_for_trial_id[trial_id] = config
+            else:
+                logger.warning(
+                    f"register_pending called for trial_id {trial_id} without passing config"
+                )
+
+    def evaluation_failed(self, trial_id: str):
+        if self._allow_duplicates and trial_id in self._config_for_trial_id:
+            # Blacklist this configuration
+            self._excl_list.add(self._config_for_trial_id[trial_id])
```
Comment (mseeger) [inline syne_tune/optimizer/schedulers/searchers/bore/bore.py:324]: Only get `trial_id` here, but need to know the config. This  is why `_config_for_trial` is needed, only if `allow_duplicates`.

File: syne_tune/optimizer/schedulers/searchers/random_grid_searcher.py [hunk 3]
```diff
@@ -136,12 +147,45 @@ def _update(self, trial_id: str, config: dict, result: dict):
...
+                self._config_for_trial_id[trial_id] = config
+            else:
+                logger.warning(
+                    f"register_pending called for trial_id {trial_id} without passing config"
+                )
+
+    def evaluation_failed(self, trial_id: str):
+        if self._allow_duplicates and trial_id in self._config_for_trial_id:
+            # Blacklist this configuration
+            self._excl_list.add(self._config_for_trial_id[trial_id])
```
Comment (wistuba) [inline syne_tune/optimizer/schedulers/searchers/random_grid_searcher.py:167]: I've seen these functions now a couple of times. Is it possible to move it up in the class hierarchy or use a utils function instead?

File: tst/schedulers/test_searchers.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        (HyperTune, True),
+        (SyncHyperband, True),
+        (DEHB, True),
+        (BOHB, True),
+        (SyncBOHB, True),
+        (BORE, False),
+        (KDE, False),
+    ],
+)
+def test_allow_duplicates_or_not(scheduler_cls, multifid):
```
Comment (wistuba) [inline tst/schedulers/test_searchers.py:58]: seems natural to move `allow_duplicates` here

File: syne_tune/optimizer/schedulers/searchers/random_grid_searcher.py (hunk 3 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/random_grid_searcher.py:167]: Yes, doing that

File: tst/schedulers/test_searchers.py (hunk 4 above)
Comment (mseeger) [reply tst/schedulers/test_searchers.py:58]: What do you mean?

Comment (wistuba) [reply tst/schedulers/test_searchers.py:58]: add another `parametrize` line with `allow_duplicates` before the function

**Comment by mseeger** [comment]
Hi @wistuba , I am still working on this. I'd like to extend the unit test to the case where all trials fail, which should filter them out even if `allow_duplicates == True`. But this does not work right now.

**Comment by mseeger** [comment]
OK, I am done with this now

File: syne_tune/optimizer/schedulers/synchronous/hyperband.py [hunk 5]
```diff
@@ -290,6 +290,9 @@ def _suggest(self, trial_id: int) -> Optional[TrialSuggestion]:
             config = self.searcher.get_config(trial_id=str(trial_id))
             if config is not None:
                 config = cast_config_values(config, self.config_space)
+                self.searcher.register_pending(
```
Comment (wistuba) [inline syne_tune/optimizer/schedulers/synchronous/hyperband.py:293]: are these changes related to the feature?

File: syne_tune/optimizer/schedulers/searchers/searcher.py [hunk 6]
```diff
@@ -450,3 +451,72 @@ def _restore_from_state(self, state: dict):
 
     def set_random_state(self, random_state: np.random.RandomState):
         self.random_state = random_state
+
+
+class FilterDuplicatesMixin:
```
Comment (wistuba) [inline syne_tune/optimizer/schedulers/searchers/searcher.py:456]: the mixin should not have functions overlapping with BaseSearcher

**Review by wistuba — CHANGES_REQUESTED** [review]
The multi-inheritance is not a good choice here. What you could do instead is make it part of the base searcher. Searchers who don't support it can simply hard code the choice of allow_duplicates to False.

File: syne_tune/optimizer/schedulers/synchronous/hyperband.py (hunk 5 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/synchronous/hyperband.py:293]: Yes, this is something that was missing and came to light because of the new test

File: syne_tune/optimizer/schedulers/searchers/searcher.py (hunk 6 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/searcher.py:456]: Yes, so this would basically be multiple inheritance then.

The problem with moving this into BaseSearcher:
- BaseSearcher is really supposed to be just an API. David would rightly reject putting code in there
- Some of the searchers do this `allow_duplicates` in a different way

Is multiple inheritance really so evil?

Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/searcher.py:456]: OK, I think what I can do, is I can integrate this code into `SearcherWithRandomSeed`. This way, the `BaseSearcher` remains clean.

Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/searcher.py:456]: Or better: make it a subclass of that one, this is best

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py [hunk 7]
```diff
@@ -85,12 +82,16 @@ def __init__(
         random_prob: Optional[float] = None,
         init_random: Optional[int] = None,
         classifier_kwargs: Optional[dict] = None,
+        allow_duplicates: Optional[bool] = None,
         **kwargs,
     ):
+        if allow_duplicates is None:
+            allow_duplicates = False
```
Comment (wistuba) [inline syne_tune/optimizer/schedulers/searchers/bore/bore.py:89]: deal with this in parent class

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py [hunk 8]
```diff
@@ -246,7 +244,8 @@ def wrapper(x):
                 f"config={config}] "
                 f"optimization time : {opt_time}"
             )
-            self._excl_list.add(config)  # Should not be suggested again
+            if not self.allow_duplicates:
+                self._excl_list.add(config)  # Should not be suggested again
```
Comment (wistuba) [inline syne_tune/optimizer/schedulers/searchers/bore/bore.py:248]: can't we delegate to parent class as well?

File: syne_tune/optimizer/schedulers/searchers/kde/kde_searcher.py [hunk 9]
```diff
@@ -96,12 +90,16 @@ def __init__(
         num_candidates: Optional[int] = None,
         bandwidth_factor: Optional[int] = None,
         random_fraction: Optional[float] = None,
+        allow_duplicates: Optional[bool] = None,
         **kwargs,
     ):
+        if allow_duplicates is None:
+            allow_duplicates = False
```
Comment (wistuba) [inline syne_tune/optimizer/schedulers/searchers/kde/kde_searcher.py:97]: same

**Review by wistuba — COMMENTED** [review]
There are few lines of code that might be suited in the newly added class.

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py (hunk 7 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/bore/bore.py:89]: OK

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py (hunk 8 above)
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/bore/bore.py:248]: I'll  try something

**Review by wistuba — DISMISSED** [review]

**Review by wistuba — APPROVED** [review]
